### ISSUE TITLE
The value of path label should stay in pattern defination

### ISSUE THREAD

**Issue by liyuntao** [vert-x3/vertx-micrometer-metrics#34]
Assuming we have such rest-style route defination:
```kotlin 
this.post("/api/v1/user/:userId").handler{ it.response().end("ok") }
this.post("/api/v1/timeline/:timelineId/comment/:commentId").handler{ it.response().end("ok") }
```

Currently when there's a request with new id hitting the server, there will be at least one new item generated.

```
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/user/1",remote="_",} 0.0
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/user/2",remote="_",} 0.0
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/user/3",remote="_",} 0.0
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/timeline/a01/comment/1",remote="_",} 0.0
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/timeline/a01/comment/2",remote="_",} 0.0
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/timeline/a02/comment/1",remote="_",} 0.0
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/timeline/a02/comment/3",remote="_",} 0.0
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/timeline/a02/comment/4",remote="_",} 0.0
```

I hope it can be like this:
```
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/registry/user/:userId",remote="_",} 0.0
vertx_http_server_requests{local="0.0.0.0:8080",path="/api/v1/timeline/:timelineId/comment/:commentId",remote="_",} 0.0
```

**Comment by jotak** [comment]
I wouldn't say the path *should* show its pattern definition, because the actual path may matter to some use case.

However you can define your own rules on labels. For instance you could do something like:

```
    MeterRegistry registry = BackendRegistries.getDefaultNow();

    registry.config().meterFilter(MeterFilter.replaceTagValues("path", path -> {
      // do some regex matching
      return modifiedPath;
    }));
```

It's even recommended to either ignore labels that would have a high cardinality such as these  dynamic paths, or "rework" them to remove the dynamic part, but we cannot assume that nobody will want the actual path.

**Comment by liyuntao** [comment]
Yeah I agree that. It maybe more reasonable if we make the path tag configurable.

The scene originated from my recent work, which is Integrating monitoring capabilities into a variety of framework we are using (vertx 3.5.x, spring-boot 2.x). In spring boot 2, there's a lib called spring-boot-actuator which also provides out-of-box monitoring capabilities using micrometer. The uri label in actuator metric items behaved the same as the one I described in title. Such kind of implementation attracted me by two points:

**1. more correspond to prometheus‘s suggested practice**
according to:
>CAUTION: Remember that every unique key-value label pair represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

**2. able to distinguish by api (with clear granularity)**
Compare to specific path, the pattern(path in route definition) would be meaningful under monitoring scenarios, basically almost all monitoring or measuring scenarios are based by route definition. While the usage of specific path sounds more like an event logging use case such as ELK stack.

---

Otherwise, the implementation detail in spring-boot-actuator project is located in [link](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/web/servlet/WebMvcTags.java#L89) 
I wonder whether it could be easily implement under vertx's current architecture(whether implement in current lib or user's custom meterFilter)

**Comment by jotak** [comment]
> However, it maybe more reasonable if we could make the path tag static or at least configurable.
How do you mean configurable ? The code snippet that I show above is a way to configure it, so the current API allows it. But I recognize that it requires some work on the user side to manipulate regexs.

I'm investigating how feasible it is to provide directly the path pattern, because right now the metrics implementations doesn't have access (as far as I know) to this information so it would probably need some interaction with other vertx modules. I will confirm that soon.

Also, note that another work is planned in order to change the default labels: see https://github.com/vert-x3/vertx-micrometer-metrics/issues/29
Basically, all potential high-cardinality labels will be opt-in, so deactivated by default. That will prevent such misuse with the defaults.

**Comment by santiagopoli** [comment]
+1 For this. Having to manually configure Regexes with all known paths seems like a burden to me.

**Comment by u6f6o** [comment]
+1 to provide a label for the origin path pattern (e.g. path="/api/v1/registry/user/:userId")

I recently wrote a micrometer wrapper for our clojure services and initially provided the fully resolved path as well. In consequence, our prometheus instances on the production system died shortly after the release and later on the migrated services as well due to OOM.

If you define paths containing for example UUIDs to identify users, events or the like, this can be a big problem, especially if you activate histograms as well. 

Providing the path pattern instead allows to define certain SLAs/alerts in your monitoring system as @liyuntao mentioned.

**Comment by jotak** [comment]
For a quick status about this issue: problem is that the metrics implementation only has access, to this day, to the `HttpClientRequest` object (from vertx-core) , whereas the static information that we'd like to have is defined and "belongs" to `RoutingContext` from vertx-web.
So we would need something like passing more context into `HttpClientRequest` so that metrics can use it.

**Comment by jotak** [comment]
@vietj any suggestion about that?

**Comment by vietj** [comment]
we could define extra callbacks in the SPI metrics that would be called by Vert.x Web ?

i.e something like `void requestRounted(R requestMetric, String route)`

this would be called by Vert.x Web or a similar framework.

**Comment by jotak** [comment]
@vietj sounds good, planning it for 4.0 ?

**Comment by vietj** [comment]
yes rather 4.0
